### ISSUE TITLE
Confusion about effective batch_size handling of horovod

### ISSUE THREAD

**Issue by sincewhenUCLA** [uber/horovod#699]
I am testing using horovod to transfer my single GPU tensorflow training code to multi-gpu one. I have made it work as in the example: tensorflow_minist.py. But I am very confused about how exactly horovod are handling the effective batch size for multi-gpu. As shown in the code snippet below which is a part of tensorflow_minist.py, the batch_size is set as 100. According to my understanding from the Docs and #241 , the effective batch size will become 100 * number_of_GPU (NG). But the following code only generate 100 images per batch inside the while loop, how would Horovod change it to 100 * NG  behind the scene? Does it copy these 100 images for NG times and pass each copy to one GPU? This part is not very clear for me. Thanks in advance!

```
training_batch_generator = train_input_generator(x_train,
                                                     y_train, batch_size=100)
    # The MonitoredTrainingSession takes care of session initialization,
    # restoring from a checkpoint, saving to a checkpoint, and closing when done
    # or an error occurs.
    with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
                                           hooks=hooks,
                                           config=config) as mon_sess:
        while not mon_sess.should_stop():
            # Run a training step synchronously.
            image_, label_ = next(training_batch_generator)
            mon_sess.run(train_op, feed_dict={image: image_, label: label_})

```

**Comment by zqlai** [comment]
Just my personal opinion.
When you train on NG GPUs, each GPU handles 100 images for each step. So for the whole training task, it handles 100*NG images per step, which is the batch_size for the whole task.
As the training data is shuffled randomly, the 100-images for each GPU are different.

**Comment by tgaddair** [comment]
@sincewhenUCLA, in Horovod, the code (your python script) is executed independently on every worker. So if you have NG workers, then each worker is going to be reading a different batch of 100 images (assuming each shuffles them randomly).

You can see how this works for yourself with a simple toy example:

```
import horovod.tensorflow as hvd
hvd.init()
print("rank: %d / %d" % (hvd.rank(), hvd.size())
```

Then if you `mpirun -np 4 ...` this script, you'll see results like:

```
rank: 0 / 4
rank: 1 / 4
rank: 2 / 4
rank: 3 / 4
```

So it's the same code, executed by multiple processes, where every process maps to a single GPU.

**Comment by sincewhenUCLA** [comment]
> Then if you `mpirun -np 4 ...` this script, you'll see results like:

Thanks for the explanation! This makes a lot sense! During the training process, it also prints NG lines of training loss when using NG workers.

In this setting, I think the training data has to be shuffled to make the Multi-GPU setting useful. Otherwise, all GPUs are training using the same batch starting from the same loss (overall loss). The resulting updated gradients will be the same as using single GPU for each step.

**Comment by LiweiPeng** [comment]
The way I distributed dataset among N GPUs is to use dataset.shard(). see this link for example. https://github.com/uber/horovod/issues/223#issuecomment-377831331
